Internal MathML tree

dpvc opened the pull request:
This is my version of the internal MathML tree structure, in Typescript.  It implements all the functionality that was in the old ElementJax (though not always in the same way).  It includes node factory code to make it possible to create these nodes (in a customizable way), and visitor functions to be able to traverse the tree (again, in a customizable way).

There is a test file in `samples/tex-nodes.js` that can be used to test the format using TeX input.  Use

    node load samples/tex-nodes.js 'tex'

where `tex` is TeX code to be parsed into the internal format.  The output will deserialized MathML with inherited attributes enclose in `[...]`, extra properties applied by the TeX input jax in `[[...]]`, and some of the important internal values enclosed in `{{...}}`.

zorkow commented on v3/MmlTree/ts/MmlNodes/mrow.ts:
```diff
@@ -0,0 +1,213 @@
+            this.getPrevClass(prev);
+            prev = null;
+            for (const child of this.childNodes) {
+                prev = child.setTeXclass(prev);
```
Why do we set TeXclass of the first element to null in this case?
This needs some extra commenting.

dpvc replied:
> Why do we set TeXclass of the first element to null in this case?

We don't.  We set the *previous* element to null, because within `\left...\right`, the first child has no previous element to affect its spacing.

In TeX, the spacing is based on two adjacent items, and `setTeXclass(prev)` isn't setting the class to be `prev`, it is setting the TeX class given that the previous item is `prev`.  There is none in this case.

If that's not what you are meaning, let me know.

zorkow commented on v3/MmlTree/ts/MmlNodes/mrow.ts:
```diff
@@ -0,0 +1,213 @@
+
+    /*
+     * @return {boolean}  This node is not considered a parent node
+     */
```
override
l

dpvc commented on v3/MmlTree/ts/Attributes.ts:
```diff
@@ -0,0 +1,191 @@
+     *                      node (not inherited or defaulted), null otherwise
+     */
+    public getExplicit(name: string) {
+        if (!this.attributes.hasOwnProperty(name)) {
```
Well, it's really an enhanced PropertyList -- one with a prototype chain that inherits from the `inherited` and `default` PropertyLists.  It acts like a PropertyList, but the properties may come from any of three places:  the explicit properties in `attributes`, or the ones in `inherited` or the ones in `defaults`.  This is why we don't have to implement the old `Get` function ourselves; object inheritance does that for us.

So to determine if an attributes has been given explicitly (rather than being inherited), we need to use `hasOwnProperty()`.

dpvc commented on v3/MmlTree/ts/Node.ts:
```diff
@@ -0,0 +1,355 @@
+ *  The basic Node and NodeClass types
+ */
+export type Node = INode;
+export type NodeClass = INodeClass;
```
OK.

dpvc commented on v3/MmlTree/ts/Node.ts:
```diff
@@ -0,0 +1,355 @@
+    /*
+     * The parent node for this one
+     */
+    public parent: Node = null;
```
I don't understand the comment.  Can you explain further what you mean?  (Note that `Node` is the same as `INode` by the definition of `Node`).

dpvc commented on v3/MmlTree/ts/Node.ts:
```diff
@@ -0,0 +1,355 @@
+    /*
+     * The NodeFactory to use to create additional nodes, as needed
+     */
+    public factory: NodeFactory = null;
```
It didn't seem necessary to add that overhead in order to prevent something we weren't going to do anyway.

dpvc commented on v3/MmlTree/ts/Node.ts:
```diff
@@ -0,0 +1,355 @@
+     * @implements {INode}
+     */
+    constructor(factory: NodeFactory, properties: PropertyList = {}, children: Node[] = []) {
+        this.factory = factory;
```
I'm not sure I understand what `public` means for the argument of `constructor()`.  I don't know what you mean by `declaration of factory field class body`.  If this is just part of the "use a getter to prevent changing the value", then I get it, but if it is something else, I don't.

dpvc commented on v3/MmlTree/ts/Node.ts:
```diff
@@ -0,0 +1,355 @@
+     */
+    public childIndex(node: Node) {
+        let i = 0;
+        for (const child of this.childNodes) {
```
I'm too used to avoiding features not available in older versions of IE.

dpvc commented on v3/MmlTree/ts/Node.ts:
```diff
@@ -0,0 +1,355 @@
+     * @override
+     */
+    public childIndex(node: Node) {
+        return null as number;
```
Yes.  Otherwise it gets a return type of `any`, and typescript (with our configuration) complains.

dpvc commented on v3/mathjax/legacy/jax/element/MmlNode.js:
```diff
@@ -0,0 +1,110 @@
+  MML.mbase.Augment({
+    toMmlNode: function (factory) {
+      var m = this.data.length;
+      var kind = this.type; if (kind === 'texatom') kind = 'TeXAtom';
```
This is a legacy file that is only used as glue to get from the old internal format to the new internal format, so this is not going to end up as part of v3.  So I'm not too worried about format, here.

dpvc commented on v3/mathjax/legacy/jax/element/MmlNode.js:
```diff
@@ -0,0 +1,110 @@
+      var kind = this.type; if (kind === 'texatom') kind = 'TeXAtom';
+      if (this.inferred) kind = 'inferredMrow';
+      var node = factory.MML[kind]();
+      for (var i = 0; i < m; i++) {
```
Again, this file is only temporary.

dpvc commented on v3/mathjax/legacy/jax/element/MmlNode.js:
```diff
@@ -0,0 +1,110 @@
+        for (var id in defaults) {
+          if (!skip[id] && !copy[id] && defaults.hasOwnProperty(id)) {
+            if (this[id] != null && this[id] !== defaults[id]) {
+              if (this.Get(id,null,1) !== this[id]) node.attributes.set(id,this[id]);
```
ditto.

dpvc commented on v3/mathjax/legacy/jax/element/MmlNode.js:
```diff
@@ -0,0 +1,110 @@
+  
+  MML.munderover.Augment({
+    toMmlNode: function (factory) {
+      var kind = (this.data[this.under] == null ? 'mover' :
```
Yes, but in the next line, not this one.

dpvc commented on v3/MmlTree/ts/Visitor.ts:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    /*
+     * Visit the tree by visiting the root node
```
I don't really know what `@override` does, or what is needed for it.  I would think that in some cases, a new comment would be needed to explain what the overridden function does (i.e., what its additional functionality is that isn't in the original).

dpvc commented:
@zorkow, OK, I've addressed all the comments, either in line above, or by changes in the code.  Some of the changes don't seem to have closed the comments above, however, but most did.

So you can do a second pass at this point.

dpvc commented:
@zorkow, thanks!

pkra commented:
:fireworks: fireworks: fireworks:
